lib/spack/docs/containers.rst

```diff
@@ -172,7 +191,65 @@ of environments:
          app: "gromacs"
          mpi: "mpich"
 
-The tables below describe the configuration options that are currently supported:
+------------------------
+Using Custom Base Images
+------------------------
+
+The ``images`` attribute in the ``container`` section has two modes of
+operation to select a proper base image for both the ``build`` and
```

Comment:
I've been reading through discussion like https://github.com/spack/spack/pull/15028#discussion_r403925148 again and I think this schema could be workable but I think these docs need to be improved. For example:

> The ``images`` attribute in the ``container`` section has two modes of operation to select a proper base image for both the ``build`` and``final`` stage.

This is confusing because this sentence specifies that there are two modes, but then only specifies the variables for one of those modes. It implies that these variables are used in both cases.

Overall I think the description should do the following

* specify that their are two modes
* specify the variables used in each mode
* include the exact yaml for each example in this section (you avoid this for os/spack because it appears in a larger example above, but that example doesn't specifically address these details so I think they should specifically be called out in this section)
* specify at a high level how spack operates in each mode
  * e.g. `in os/spack mode, the user specifies a Spack version and OS. Spack has pre-constructed images for the OS with that Spack version pre-installed; Spack only supports a limited number of versions/operating systems in this way, so in this case the OS used is constrained. Spack performs the build on an image of the OS with Spack, then installs files to a corresponding base image of the OS (i.e. without Spack installed).`
  * e.g. `in build/final mode the user can specify any image to perform the Spack build on and any image to install the resulting files to. In this mode the user is responsible for creating the images and managing all of their details (e.g. that the desired Spack version is provided).`

Reply:
@scheibelp I reworked this section completely, let me know what you think of the docs now. If this seems the right way to proceed I'd be happy to add a Boolean (as we discussed off-line):
```yaml
container:
  images:
    advanced: True
    build: myrepo/myimage:latest
    final: ubuntu:18.04
```
to explicitly activate the advanced mode.

Reply:
@scheibelp ping